rpcs3/Loader/ISO.cpp

```diff
@@ -0,0 +1,600 @@
+
+void iso_form_hierarchy(fs::file& file, iso_fs_node& node,
+	bool use_ucs2_decoding = false, std::string parent_path = "")
+{
+	if (!node.metadata.is_directory) return;
+
+	std::vector<int> multi_extent_node_indices;
+
+	// assuming the directory spans a single extent
+	const auto& directory_extent = node.metadata.extents[0];
+
+	file.seek(directory_extent.start * ISO_BLOCK_SIZE);
+
+	u64 end_pos = directory_extent.size + (directory_extent.start * ISO_BLOCK_SIZE);
+
+	while(file.pos() < end_pos)
+	{
+		auto entry = iso_read_directory_entry(file, use_ucs2_decoding);
+		if (!entry)
+		{
+			float block_size = ISO_BLOCK_SIZE;
+			float t = std::floor(file.pos() / block_size);
+			u64 new_sector = t+1;
+			file.seek(new_sector * ISO_BLOCK_SIZE);
+			continue;
+		}
+
+		bool extent_added = false;
+
+		// find previous extent and merge into it, otherwise we push this node's index
+		for (int index : multi_extent_node_indices)
+		{
+			auto& selected_node = node.children.at(index);
+			if (selected_node->metadata.name.compare(entry->name) == 0)
+			{
+				// merge into selected_node
+				selected_node->metadata.extents.push_back(entry->extents[0]);
+
+				extent_added = true;
+			}
+		}
+
+		if (extent_added) continue;
+
+		if (entry->has_multiple_extents)
+		{
+			// haven't pushed entry to node.children yet so node.children::size() == entry_index
+			multi_extent_node_indices.push_back(node.children.size());
+		}
+
+		node.children.push_back(std::make_unique<iso_fs_node>(iso_fs_node{
+			.metadata = *entry
+		}));
+	}
+
+	for (auto& child_node : node.children)
+	{
+		if (child_node->metadata.name != "." && child_node->metadata.name != "..")
+		{
+			iso_form_hierarchy(file, *child_node, use_ucs2_decoding, parent_path + "/" + node.metadata.name);
+		}
+	}
+}
+
+u64 iso_fs_metadata::size() const
+{
+	u64 total_size = 0;
+	for (const auto& extent : extents)
+	{
+		total_size += extent.size;
+	}
+
+	return total_size;
+}
+
+iso_archive::iso_archive(const std::string& path)
+{
+	m_path = path;
+	m_file = fs::file(path);
+
+	if (!is_file_iso(m_file))
+	{
+		// not iso... TODO: throw something??
+		return;
+	}
+
+	u8 descriptor_type = -2;
+	bool use_ucs2_decoding = false;
+
+	do
+	{
+		auto descriptor_start = m_file.pos();
+
+		descriptor_type = m_file.read<u8>();
+
+		// 1 = primary vol descriptor, 2 = joliet SVD
+		if (descriptor_type == 1 || descriptor_type == 2)
+		{
+			use_ucs2_decoding = descriptor_type == 2;
+
+			// skip the rest of descriptor's data
+			m_file.seek(155, fs::seek_cur);
+
+			m_root = iso_fs_node
+			{
+				.metadata = iso_read_directory_entry(m_file, use_ucs2_decoding).value(),
+			};
+
+			m_file.seek(descriptor_start);
+		}
+
+		m_file.seek(descriptor_start + ISO_BLOCK_SIZE);
+	}
+	while(descriptor_type != 255);
+
+	iso_form_hierarchy(m_file, m_root, use_ucs2_decoding);
+}
+
+iso_fs_node* iso_archive::retrieve(const std::string& passed_path)
+{
+	std::string path = std::filesystem::path(passed_path).lexically_normal().string();
```

Comment:
Well techncially, this is a bug. Real filesystem when it sees something like "/directory_exist/does_not_exist_directory/../" it fails.
You need to support "/.." properly.